### ISSUE TITLE
Crash when clearing app data with settings menu open

### ISSUE THREAD

**Issue by solonovamax** [nextcloud/android#11064]
### Cause of error
You can cause a crash by clearing the Nextcloud app's data with the settings menu open.

Steps to reproduce:
1. Open settings menu in Nextcloud app
2. Open android settings menu
3. Navigate to android apps menu
4. Select Nextcloud app
5. Select "Storage"
6. Select "Clear data"
7. Return to Nextcloud app. The settings menu is still open.
8. Exit out of the settings menu. Crash.

#### Stack Trace
```java
Exception in thread "main" java.lang.RuntimeException
    at com.owncloud.android.ui.activities.ActivitiesActivity$$ExternalSyntheticLambda0.get(Unknown Source:2)
    at com.nextcloud.java.util.Optional.orElseThrow(Optional.java:306)
    at com.owncloud.android.ui.activity.FileDisplayActivity$8.run(FileDisplayActivity.java:2026)
    at android.os.Handler.handleCallback(Handler.java:883)
    at android.os.Handler.dispatchMessage(Handler.java:100)
    at android.os.Looper.loop(Looper.java:214)
    at android.app.ActivityThread.main(ActivityThread.java:7615)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:964)

```

### App information
* ID: `com.nextcloud.client`
* Version: `30220390`
* Build flavor: `gplay`

### Device information
* Brand: `lge`
* Device: `judyln`
* Model: `LM-G710`
* Id: `QKQ1.191222.002`
* Product: `judyln_lao_com`

### Firmware
* SDK: `29`
* Release: `10`
* Incremental: `8450100201101`

**Comment by ezaquarii** [comment]
@AlvaroBrey I don't think this is a real bug. This is never going to work.

The crash is caused by clearing the application data, which **removes the user from account manager at runtime**.

When application data has been deleted by the operating system *at runtime* - the effect is pretty much impossible to predict. I think Android should shut down the application when data is cleared.

The fix would require fixing *every* activity. Fortunately, we do have a single point of responsibility that gives us a chance: `SessionMixin.onStart()`. This should give us an opportunity to shut down any activity before it tries to access deleted data.
It probably won't fix all edge cases, as not all activities are using it.

TBH, this is such an odd use case that I'd `wontfix` it.

**Comment by AlvaroBrey** [comment]
> The crash is caused by clearing the application data, which **removes the user from account manager at runtime**.
> 
> When application data is deleted by the operating system _at runtime_ - the effect is pretty much impossible to predict. I think Android should shut down the application when data is cleared.
> 
> The fix would require fixing _every_ activity. Fortunately, we do have a single point of responsibility that gives us a chance: `SessionMixin.onStart()`. This should give us an opportunity to shut down any activity before it tries to access deleted data. It probably won't fix all edge cases, as not all activities are using it.
> 
> TBH, this is such an odd use case that I'd `wontfix` it.

We have a custom "manage space" activity, we're not using the system one. I believe the point of this activity is to delete only settings and database, but not the downloaded files. The problem is that when doing this, the app is not correctly killed. My previous idea was to kill the app after clearing data [1],  but it isn't working properly.

Honestly, at this point I'd rather remove the custom "Manage space" activity and just let the system clear it properly and delete the files. I don't even think not deleting the files is a desirable behaviour: if you clear data and then log in with another account, the downloaded files from the previous setup will never be deleted and will just be there using space.
What do you think, @tobiasKaminsky ?

[1] https://github.com/nextcloud/android/blob/719b0bb9a69ce73e0d3399fd2ebc4f14e64d46fe/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L118-L119

**Comment by tobiasKaminsky** [comment]
https://github.com/nextcloud/android/blob/e57176f009b5897e2000b089c033ac25ee85a62a/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L78

Manage Space clease data, but remains passcode.

Deleting app cache/file via Android also kills passcode, thus we have this custome one.

**Comment by AlvaroBrey** [comment]
> https://github.com/nextcloud/android/blob/e57176f009b5897e2000b089c033ac25ee85a62a/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L78
> 
> Manage Space clease data, but remains passcode.
> 
> Deleting app cache/file via Android also kills passcode, thus we have this custome one.

But what is the point of keeping the passcode if we delete all the other data?

**Comment by tobiasKaminsky** [comment]
No, this does not remove account, and thus passcode must remain.
It is "only" to free up local space.

Remove account is entirely removing everything, including all settings/passcode.

**Comment by AlvaroBrey** [comment]
> No, this does not remove account, and thus passcode must remain. It is "only" to free up local space.
> 
> Remove account is entirely removing everything, including all settings/passcode.

We might be talking about different things here? ManageSpaceActivity definitely [deletes the entire database and SharedPreferences](url) (except the lock, which it restores afterwards). In fact it [deletes everything under the app's private directory](https://github.com/nextcloud/android/blob/5d4b7a6d04d28402d4865179f2c7cc8690e08222/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L123-L123). And it also [deletes accounts in AccountManager as well](https://github.com/nextcloud/android/blob/5d4b7a6d04d28402d4865179f2c7cc8690e08222/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L105-L105).

**Comment by solonovamax** [comment]
> @AlvaroBrey I don't think this is a real bug. This is never going to work.
> 
> The crash is caused by clearing the application data, which **removes the user from account manager at runtime**.

imo, it is a bug.
Specifically *because* it is never going to work.

It should return the use to the "login" screen if the account cannot be properly located/logged in to.

**Comment by solonovamax** [comment]
> No, this does not remove account, and thus passcode must remain. It is "only" to free up local space.
> 
> Remove account is entirely removing everything, including all settings/passcode.

Clearing app data absolutely removes the accounts. It logs you out of all accounts, and (essentially) reverts the nextcloud app to a "clean" install. (Or at least, that's what it *should* do, imo)

**Comment by AlvaroBrey** [comment]
> > No, this does not remove account, and thus passcode must remain. It is "only" to free up local space.
> > Remove account is entirely removing everything, including all settings/passcode.
> 
> We might be talking about different things here? ManageSpaceActivity definitely [deletes the entire SharedPreferences](https://github.com/nextcloud/android/blob/5d4b7a6d04d28402d4865179f2c7cc8690e08222/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L92-L92) (except the lock, which it restores afterwards) and database. In fact it [deletes everything under the app's private directory](https://github.com/nextcloud/android/blob/5d4b7a6d04d28402d4865179f2c7cc8690e08222/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L123-L123). And it also [deletes accounts in AccountManager as well](https://github.com/nextcloud/android/blob/5d4b7a6d04d28402d4865179f2c7cc8690e08222/app/src/main/java/com/owncloud/android/ui/activity/ManageSpaceActivity.kt#L105-L105).
> 
> Also check the conversation in #9751, which is where we added the "delete accounts" step.
> 
> In short, as it's currently implemented, the ManageSpaceActivity deletes everything except the passcode, and any downloaded files that may be outside the app's private directory. This is why I am proposing to remove it, as it offers very little benefit over the default Android "clear data", and causes many more headaches.

Ping @tobiasKaminsky

**Comment by joshtrichards** [comment]
Fixed by #11518
